## LAB-5
### Name: Ramya Chandran
### StudenID: 8895789

In [80]:
# Import the required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score, mean_absolute_error
from sklearn.pipeline import Pipeline
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import cross_val_score, cross_val_predict

1. Utilize the diabetes dataset from lab 4. Perform cross-validation on nine polynomial models, ranging from degree 0 to 8. (2 points)


In [81]:
diabetes = load_diabetes()

X, y = load_diabetes(as_frame=True, scaled=False, return_X_y=True)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X.head()

age  sex   bmi     bp     s1     s2    s3   s4      s5    s6
0  59.0  2.0  32.1  101.0  157.0   93.2  38.0  4.0  4.8598  87.0
1  48.0  1.0  21.6   87.0  183.0  103.2  70.0  3.0  3.8918  69.0
2  72.0  2.0  30.5   93.0  156.0   93.6  41.0  4.0  4.6728  85.0
3  24.0  1.0  25.3   84.0  198.0  131.4  40.0  5.0  4.8903  89.0
4  50.0  1.0  23.0  101.0  192.0  125.4  52.0  4.0  4.2905  80.0

In [82]:

# Define the degrees of the polynomial models to consider
degrees = np.arange(9)
R2_score = []
MAE_score = []

# Perform cross-validation for each degree
for degree in degrees:
    # Create a polynomial feature transformer
    poly_model = make_pipeline(PolynomialFeatures(degree),LinearRegression())

    # Perform cross-validation
    scores = cross_val_score(poly_model,X,y,cv=5,scoring='neg_mean_absolute_error')
    R2_scores = cross_val_score(poly_model,X,y,cv=5,scoring='r2')
    MAE_score.append(-scores)
    R2_score.append(R2_scores)
   
    

In [83]:
print("MAE value of 5 cv model of 9 polynomial degree is \n",np.round(MAE_score, decimals=4))
print("R2 value of 5 cv model of 9 polynomial degree is \n",np.round(R2_score, decimals=4))


MAE value of 5 cv model of 9 polynomial degree is 
 [[  60.95   68.32   67.14   63.26   70.55]
 [  43.03   44.8    48.16   43.01   42.39]
 [  45.59   82.05   48.46   49.74   79.49]
 [ 167.78  215.42  315.7   485.77  525.6 ]
 [ 436.36  599.17  843.03  838.29  569.45]
 [ 512.6   625.03  556.68  484.48  636.19]
 [ 594.3  1104.32  684.58  580.88  748.95]
 [ 728.54 1805.37  930.78  790.9   907.82]
 [ 890.66 2854.91 1371.22 1150.4  1111.11]]
R2 value of 5 cv model of 9 polynomial degree is 
 [[    -0.1      -0.03     -0.       -0.01     -0.  ]
 [     0.43      0.52      0.48      0.43      0.55]
 [     0.35     -1.32      0.45      0.17     -0.89]
 [   -16.29    -26.93    -54.42   -592.67   -326.79]
 [  -171.39   -744.07  -1088.42   -729.54   -122.  ]
 [  -365.26  -1176.43   -184.92   -138.07   -319.6 ]
 [  -496.25  -6955.6    -362.28   -221.9    -439.45]
 [  -766.   -24562.84  -1170.91   -494.38   -660.34]
 [ -1272.65 -72123.06  -4411.26  -1512.35  -1061.96]]


2. Construct a table summarizing the cross-validation results. Each model should have a separate row in the table. Include the R-Squared and Mean Absolute Error (MAE) metrics for each model. Calculate the mean value and standard deviation of these metrics from the cross-validation. Include both values. (2 points)


In [84]:
# Define the degrees of the polynomial models to consider
degrees = np.arange(9)
R2_score = []
MAE_score = []

# Perform cross-validation for each degree
for degree in degrees:
    # Create a polynomial feature transformer
    poly_model = make_pipeline(PolynomialFeatures(degree),LinearRegression())

    # Perform cross-validation
    scores = cross_val_score(poly_model,X,y,cv=5,scoring='neg_mean_absolute_error').mean()
    R2_scores = cross_val_score(poly_model,X,y,cv=5,scoring='r2').mean()
    MAE_score.append(-scores)
    R2_score.append(R2_scores)
    
# Print the final_df with mean value of all model R-Squared and MAE values
final_df = pd.DataFrame({"Degree":degrees, "R2_Score":R2_score, "MAE_Score":MAE_score})
print("The final table with R2 and MAE values are : \n", final_df)

The final table with R2 and MAE values are : 
    Degree      R2_Score    MAE_Score
0       0     -0.027506    66.045624
1       1      0.482316    44.276499
2       2     -0.249574    61.065185
3       3   -203.418055   342.051797
4       4   -571.083108   657.260477
5       5   -436.856887   562.993636
6       6  -1695.096515   742.606516
7       7  -5530.894075  1032.681725
8       8 -16076.255117  1475.658532


In [85]:

# Calculate mean and standard deviation of the metrics
print("Mean R-Squared : ", final_df['R2_Score'].mean())
print("Std R-Squared : ",final_df['R2_Score'].std())
print("Mean MAE : ", final_df['MAE_Score'].mean())
print("Std MAE : ", final_df['MAE_Score'].std())

Mean R-Squared :  -2723.710946703083
Std R-Squared :  5311.556589181047
Mean MAE :  553.8488878202126
Std MAE :  488.71397067388347


3. Identification of the Best Model: Identify the model that exhibits the highest performance based on the R-Squared and MAE metrics. Provide an explanation for choosing this specific model. (2 points)



Based on the final summary of R2 and MAE values.
* The model with highest R2 predicts better and has more accuracy and that value here would be 0.48 corresponding to degree 1
* The model with least MAE is better since the error of prediction is less so the value is 44.27 corresponding to degree 1
* Based on both MAE and R2 model with degree 1 is the best model.